### PR TITLE
Include non-redundant separators in `Hash` for `Path`

### LINKED ISSUE #127254: `Hash` implementation for `Path` ignores path separators
I tried [this code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=467c0f7aedcfbfb484c9ca1ec65ff406):

```rust
use std::path::PathBuf;
use std::hash::{DefaultHasher, Hasher, Hash};

fn main() {
    let x = PathBuf::from("hello/world");
    let y = PathBuf::from("helloworld");
    
    let mut hasher = DefaultHasher::new();
    x.hash(&mut hasher);
    println!("{}", hasher.finish());

    let mut hasher = DefaultHasher::new();
    y.hash(&mut hasher);
    println!("{}", hasher.finish());
}
```

I expected to see this happen: Different hashes since these represent different paths.

Instead, this happened: The hashes are equal.

This makes sense for redundant path separators (e.g. `//`), but not for _all_ path separators.

----

I added a test case to the standard library — which fails.

```diff
diff --git a/library/std/src/path/tests.rs b/library/std/src/path/tests.rs
index 92702b395df..ed945a764c5 100644
--- a/library/std/src/path/tests.rs
+++ b/library/std/src/path/tests.rs
@@ -1566,6 +1566,13 @@ macro_rules! tc (
     relative_from: Some("bar")
     );
 
+    tc!("foo/bar", "foobar",
+    eq: false,
+    starts_with: false,
+    ends_with: false,
+    relative_from: None
+    );
+
     tc!("foo/bar/baz", "foo/bar",
     eq: false,
     starts_with: true,
```

```
---- path::tests::test_compare stdout ----
thread 'path::tests::test_compare' panicked at library/std/src/path/tests.rs:1569:5:
"foo/bar" == "foobar", expected false, got 4111386826608627770 and 4111386826608627770
```

The relevant code is probably around https://github.com/rust-lang/rust/blob/6292b2af620dbd771ebb687c3a93c69ba8f97268/library/std/src/path.rs#L3112-L3114 which last changed in https://github.com/rust-lang/rust/commit/45082b077b4991361f451701d0a9467ce123acdf. If I revert that commit, the error persists.  It looks like the relevant change is https://github.com/rust-lang/rust/commit/a083dd653af0f7f46ba6058ab51e1f9d6a2aca7d, if I revert that commit the test passes. This was the solution I opted for downstream (in https://github.com/astral-sh/ruff/pull/12159) before performing this investigation.

### Meta

I tested this on stable (1.79.0), nightly (1.81.0), and on `master` (6292b2af620dbd771ebb687c3a93c69ba8f97268).

Discovered via https://github.com/astral-sh/ruff/issues/12158

### PR DESCRIPTION
**Pull request by zanieb**
Closes https://github.com/rust-lang/rust/issues/127254, please see the issue for the details as I cover a lot there.

I'm not confident this is the clearest way to fix this, I'd appreciate some feedback.

### CODE REVIEW

**Comment by zanieb** [comment]
cc @the8472 as I believe the regression was from your commit and you look quite active still.

**Comment by zanieb** [comment]
Ah I can reproduce that test failure locally — I was only testing `test_compare`. I'll investigate a fix.

File: library/std/src/path/tests.rs [hunk 1]
```diff
@@ -1559,13 +1573,34 @@ pub fn test_compare() {
...
     ends_with: false,
     relative_from: Some("bar")
     );
 
+    tc!("foo/bar", "foobar",
+    eq: false,
+    starts_with: false,
+    ends_with: false,
+    relative_from: None
+    );
```
Comment (zanieb) [inline library/std/src/path/tests.rs:1602]: This is the test case reported in the issue.

**Comment by the8472** [comment]
I'll take a look and will have to run some benchmarks to see the perf impact of this change, since that loop is fairly sensitive.

**Comment by the8472** [comment]
Yeah, this regresses hashset performance

```
this PR: 
    path::tests::bench_path_hashset                   342.01/iter    +/- 8.37
    path::tests::bench_path_hashset_miss              139.03/iter    +/- 8.95

master:
    path::tests::bench_path_hashset                   244.21/iter    +/- 6.28
    path::tests::bench_path_hashset_miss              100.68/iter    +/- 9.11
```

I'll see if I can find an alternative that satisfies the added tests.

**Comment by the8472** [comment]
I took your tests and added add a more lightweight fix in #127297

**Comment by zanieb** [comment]
https://github.com/rust-lang/rust/pull/127297 seems like a much approach

**Pull request closed without merging**